File: app/src/main/java/com/example/muse/ai/VoiceAiAssistant.kt
package com.example.muse.ai

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.speech.RecognitionListener
import android.speech.RecognizerIntent
import android.speech.SpeechRecognizer
import android.util.Log
import com.example.muse.model.VoiceAssistantState
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import java.util.Locale
import java.util.regex.Pattern

sealed class VoiceAction {
    data class SetVolume(val targetVolume: Int) : VoiceAction()
    data class AdjustVolumeRelative(val delta: Int) : VoiceAction()
    data object Mute : VoiceAction()
    data object Unmute : VoiceAction()
    data object Play : VoiceAction()
    data object Pause : VoiceAction()
    data object NextSong : VoiceAction()
    data object PreviousSong : VoiceAction()
    data object RestartSong : VoiceAction()
    data class SearchAndPlay(val query: String) : VoiceAction()
    data class SetSpatialAudio(val enabled: Boolean) : VoiceAction()
    data class SetDolbyAtmos(val enabled: Boolean) : VoiceAction()
    data object OpenStemExtractor : VoiceAction()
    data object OpenAudioTrimmer : VoiceAction()
    data object OpenLyricsChat : VoiceAction()
    data class GeneralInfo(val response: String) : VoiceAction()
}

class VoiceAiAssistant(
    private val context: Context,
    private var onActionDispatched: ((VoiceAction) -> Unit)? = null
) {
    private val TAG = "VoiceAiAssistant"
    private var speechRecognizer: SpeechRecognizer? = null
    private val _state = MutableStateFlow(VoiceAssistantState())
    val state: StateFlow<VoiceAssistantState> = _state.asStateFlow()

    init {
        try {
            if (SpeechRecognizer.isRecognitionAvailable(context)) {
                speechRecognizer = SpeechRecognizer.createSpeechRecognizer(context).apply {
                    setRecognitionListener(createRecognitionListener())
                }
            }
        } catch (e: Exception) {
            Log.w(TAG, "SpeechRecognizer initialization warning", e)
        }
    }

    fun startListening(callback: ((VoiceAction) -> Unit)? = null) {
        if (callback != null) {
            this.onActionDispatched = callback
        }
        _state.value = _state.value.copy(
            isListening = true,
            recognizedText = "",
            assistantFeedback = "Listening for 'Hey Muse' or your command..."
        )

        try {
            val intent = Intent(RecognizerIntent.ACTION_RECOGNIZE_SPEECH).apply {
                putExtra(RecognizerIntent.EXTRA_LANGUAGE_MODEL, RecognizerIntent.LANGUAGE_MODEL_FREE_FORM)
                putExtra(RecognizerIntent.EXTRA_LANGUAGE, Locale.getDefault())
                putExtra(RecognizerIntent.EXTRA_PARTIAL_RESULTS, true)
                putExtra(RecognizerIntent.EXTRA_MAX_RESULTS, 3)
            }
            speechRecognizer?.startListening(intent)
        } catch (e: Exception) {
            Log.e(TAG, "Failed to start speech recognizer", e)
            _state.value = _state.value.copy(
                assistantFeedback = "Mic ready. Tap command or speak 'Hey Muse...'"
            )
        }
    }

    fun stopListening() {
        try {
            speechRecognizer?.stopListening()
        } catch (e: Exception) {
            Log.e(TAG, "Error stopping recognizer", e)
        }
        _state.value = _state.value.copy(isListening = false)
    }

    fun parseCommand(rawInput: String): VoiceAction {
        val cleanInput = rawInput.trim().lowercase(Locale.ROOT)
        val normalized = cleanInput
            .replace("hey muse", "")
            .replace("ok muse", "")
            .replace("muse", "")
            .trim()

        val textToEvaluate = if (normalized.isNotEmpty()) normalized else cleanInput

        // 1. Explicit Volume 0 - 100
        val explicitVolumeMatch = Pattern.compile(".*?(?:volume|sound|level)\\s*(?:to|at|is|=)?\\s*(\\d{1,3})%?.*").matcher(textToEvaluate)
        if (explicitVolumeMatch.find()) {
            val volVal = explicitVolumeMatch.group(1)?.toIntOrNull()
            if (volVal != null) {
                return VoiceAction.SetVolume(volVal.coerceIn(0, 100))
            }
        }

        // Relative Volume
        if (textToEvaluate.contains("decrease volume") || textToEvaluate.contains("volume down") || textToEvaluate.contains("lower sound") || textToEvaluate.contains("quieter")) {
            val numberMatch = Pattern.compile(".*?(\\d{1,2}).*").matcher(textToEvaluate)
            val delta = if (numberMatch.find()) -(numberMatch.group(1)?.toIntOrNull() ?: 15) else -15
            return VoiceAction.AdjustVolumeRelative(delta)
        }

        if (textToEvaluate.contains("increase volume") || textToEvaluate.contains("volume up") || textToEvaluate.contains("raise sound") || textToEvaluate.contains("louder")) {
            val numberMatch = Pattern.compile(".*?(\\d{1,2}).*").matcher(textToEvaluate)
            val delta = if (numberMatch.find()) (numberMatch.group(1)?.toIntOrNull() ?: 15) else 15
            return VoiceAction.AdjustVolumeRelative(delta)
        }

        // Mute / Unmute
        if (textToEvaluate.contains("unmute") || textToEvaluate.contains("sound on")) {
            return VoiceAction.Unmute
        }
        if (textToEvaluate.contains("mute") || textToEvaluate.contains("silence")) {
            return VoiceAction.Mute
        }

        // Playback Navigation
        if (textToEvaluate.contains("next song") || textToEvaluate.contains("next track") || textToEvaluate.contains("skip")) {
            return VoiceAction.NextSong
        }
        if (textToEvaluate.contains("previous song") || textToEvaluate.contains("previous track") || textToEvaluate.contains("go back")) {
            return VoiceAction.PreviousSong
        }
        if (textToEvaluate.contains("restart") || textToEvaluate.contains("replay")) {
            return VoiceAction.RestartSong
        }
        if (textToEvaluate.contains("pause") || textToEvaluate.contains("stop")) {
            return VoiceAction.Pause
        }
        if (textToEvaluate == "play" || textToEvaluate == "resume" || textToEvaluate.contains("start playing")) {
            return VoiceAction.Play
        }

        // Search & Play
        if (textToEvaluate.startsWith("play ") || textToEvaluate.contains("play track") || textToEvaluate.contains("search for")) {
            val query = textToEvaluate
                .replace("play track", "")
                .replace("search for", "")
                .replace("play", "")
                .trim()
            return VoiceAction.SearchAndPlay(query)
        }

        // Spatial Audio & Dolby
        if (textToEvaluate.contains("spatial audio") || textToEvaluate.contains("3d audio")) {
            val enable = !textToEvaluate.contains("off") && !textToEvaluate.contains("disable")
            return VoiceAction.SetSpatialAudio(enable)
        }
        if (textToEvaluate.contains("dolby atmos") || textToEvaluate.contains("dolby audio")) {
            val enable = !textToEvaluate.contains("off") && !textToEvaluate.contains("disable")
            return VoiceAction.SetDolbyAtmos(enable)
        }

        // Utilities
        if (textToEvaluate.contains("bgm") || textToEvaluate.contains("stem") || textToEvaluate.contains("extract layer")) {
            return VoiceAction.OpenStemExtractor
        }
        if (textToEvaluate.contains("trim") || textToEvaluate.contains("crop") || textToEvaluate.contains("ringtone")) {
            return VoiceAction.OpenAudioTrimmer
        }
        if (textToEvaluate.contains("lyrics") || textToEvaluate.contains("words") || textToEvaluate.contains("notes")) {
            return VoiceAction.OpenLyricsChat
        }

        return VoiceAction.GeneralInfo("Executing command: $rawInput")
    }

    fun setManualInputFeedback(commandText: String, action: VoiceAction) {
        val feedback = when (action) {
            is VoiceAction.SetVolume -> "Adjusted volume to ${action.targetVolume}%."
            is VoiceAction.AdjustVolumeRelative -> if (action.delta > 0) "Increasing volume..." else "Decreasing volume..."
            is VoiceAction.Mute -> "Audio muted."
            is VoiceAction.Unmute -> "Audio unmuted."
            is VoiceAction.NextSong -> "Skipping to next track..."
            is VoiceAction.PreviousSong -> "Skipping to previous track..."
            is VoiceAction.RestartSong -> "Restarting track..."
            is VoiceAction.Play -> "Resuming audio playback..."
            is VoiceAction.Pause -> "Playback paused."
            is VoiceAction.SearchAndPlay -> "Searching & playing '${action.query}'..."
            is VoiceAction.SetSpatialAudio -> if (action.enabled) "Spatial 3D Audio enabled." else "Spatial Audio bypassed."
            is VoiceAction.SetDolbyAtmos -> if (action.enabled) "Dolby Atmos engine active." else "Dolby Atmos bypassed."
            is VoiceAction.OpenStemExtractor -> "Opening BGM Multi-Stem Segment Explorer..."
            is VoiceAction.OpenAudioTrimmer -> "Opening Audio Trimmer Studio..."
            is VoiceAction.OpenLyricsChat -> "Opening Muse Lyrics Intelligence..."
            is VoiceAction.GeneralInfo -> action.response
        }

        _state.value = _state.value.copy(
            recognizedText = commandText,
            assistantFeedback = feedback,
            isListening = false
        )
    }

    fun parseAndExecute(rawInput: String) {
        val action = parseCommand(rawInput)
        setManualInputFeedback(rawInput, action)
        onActionDispatched?.invoke(action)
    }

    private fun createRecognitionListener() = object : RecognitionListener {
        override fun onReadyForSpeech(params: Bundle?) {
            _state.value = _state.value.copy(assistantFeedback = "Hey Muse is listening...")
        }

        override fun onBeginningOfSpeech() {}
        override fun onRmsChanged(rmsdB: Float) {}
        override fun onBufferReceived(buffer: ByteArray?) {}
        override fun onEndOfSpeech() {
            _state.value = _state.value.copy(isListening = false)
        }

        override fun onError(error: Int) {
            _state.value = _state.value.copy(
                isListening = false,
                assistantFeedback = "Tap the orb to speak or pick a command chip below."
            )
        }

        override fun onResults(results: Bundle?) {
            val matches = results?.getStringArrayList(SpeechRecognizer.RESULTS_RECOGNITION)
            val topResult = matches?.firstOrNull() ?: ""
            if (topResult.isNotBlank()) {
                parseAndExecute(topResult)
            }
        }

        override fun onPartialResults(partialResults: Bundle?) {
            val matches = partialResults?.getStringArrayList(SpeechRecognizer.RESULTS_RECOGNITION)
            val partial = matches?.firstOrNull() ?: ""
            if (partial.isNotBlank()) {
                _state.value = _state.value.copy(recognizedText = partial)
            }
        }

        override fun onEvent(eventType: Int, params: Bundle?) {}
    }

    fun destroy() {
        try {
            speechRecognizer?.destroy()
            speechRecognizer = null
        } catch (e: Exception) {
            Log.e(TAG, "Error destroying speech recognizer", e)
        }
    }
}

File: src/ai/voiceAiAssistant.ts
import { VoiceAction, VoiceAssistantState } from '../types';

// SpeechRecognition type declarations for browser support
interface IWindow extends Window {
  SpeechRecognition?: any;
  webkitSpeechRecognition?: any;
}

export class VoiceAiAssistant {
  private recognition: any = null;
  private isListening: boolean = false;
  private onActionDispatched?: (action: VoiceAction) => void;
  private onStateUpdated?: (state: VoiceAssistantState) => void;
  private isSpeechOutputEnabled: boolean = true;
  private isWakeWordActive: boolean = false;

  private state: VoiceAssistantState = {
    isListening: false,
    isAlwaysListeningEnabled: false,
    isSpeaking: false,
    recognizedText: '',
    assistantFeedback: "Listening for 'Hey Muse' or your command...",
    lastAction: null,
    confidence: 0.95
  };

  constructor() {
    this.initSpeechRecognition();
  }

  private initSpeechRecognition() {
    const win = typeof window !== 'undefined' ? (window as unknown as IWindow) : null;
    const SpeechRecognitionClass = win?.SpeechRecognition || win?.webkitSpeechRecognition;

    if (SpeechRecognitionClass) {
      try {
        this.recognition = new SpeechRecognitionClass();
        this.recognition.continuous = false;
        this.recognition.interimResults = true;
        this.recognition.lang = 'en-US';

        this.recognition.onstart = () => {
          this.isListening = true;
          this.updateState({
            isListening: true,
            assistantFeedback: "Hey Muse is listening..."
          });
        };

        this.recognition.onresult = (event: any) => {
          const results = event.results;
          if (!results || results.length === 0) return;

          const transcript = results[0][0].transcript;
          const isFinal = results[0].isFinal;

          this.updateState({ recognizedText: transcript });

          if (isFinal) {
            this.parseAndExecute(transcript);
          }
        };

        this.recognition.onerror = (event: any) => {
          console.warn('Speech recognition notice:', event.error);
          this.isListening = false;
          this.updateState({
            isListening: false,
            assistantFeedback: event.error === 'not-allowed'
              ? 'Microphone access blocked. Click chips or type commands below.'
              : 'Tap microphone orb or pick a quick command chip.'
          });
        };

        this.recognition.onend = () => {
          this.isListening = false;
          this.updateState({ isListening: false });

          // Auto-restart if wake-word / always-listening mode is active
          if (this.state.isAlwaysListeningEnabled && this.isWakeWordActive) {
            setTimeout(() => {
              if (this.state.isAlwaysListeningEnabled) {
                this.startListening();
              }
            }, 800);
          }
        };
      } catch (err) {
        console.warn('Speech recognition could not be initialized:', err);
      }
    }
  }

  public setCallbacks(
    onAction: (action: VoiceAction) => void,
    onState: (state: VoiceAssistantState) => void
  ) {
    this.onActionDispatched = onAction;
    this.onStateUpdated = onState;
    onState(this.state);
  }

  public setSpeechOutput(enabled: boolean) {
    this.isSpeechOutputEnabled = enabled;
  }

  public getSpeechOutput(): boolean {
    return this.isSpeechOutputEnabled;
  }

  public toggleAlwaysListening(enabled: boolean) {
    this.isWakeWordActive = enabled;
    this.updateState({ isAlwaysListeningEnabled: enabled });
    if (enabled && !this.isListening) {
      this.startListening();
    } else if (!enabled && this.isListening) {
      this.stopListening();
    }
  }

  public startListening() {
    if (this.recognition) {
      try {
        this.recognition.abort();
      } catch (e) {
        // ignore
      }
      try {
        this.recognition.start();
        this.isListening = true;
      } catch (e) {
        this.updateState({ isListening: true });
      }
    } else {
      this.updateState({
        isListening: true,
        recognizedText: '',
        assistantFeedback: "Web Speech API ready. Speak command or tap chip below..."
      });
      // Emulate listening window timeout if no native mic recognition
      setTimeout(() => {
        if (this.state.isListening && !this.state.recognizedText) {
          this.updateState({
            isListening: false,
            assistantFeedback: "Tap microphone or choose a voice command chip below."
          });
        }
      }, 4000);
    }
  }

  public stopListening() {
    if (this.recognition) {
      try {
        this.recognition.stop();
      } catch (e) {
        // ignore
      }
    }
    this.isListening = false;
    this.updateState({ isListening: false });
  }

  /**
   * Spoken speech output via Web Speech Synthesis API
   */
  public speak(text: string) {
    if (!this.isSpeechOutputEnabled) return;
    if (typeof window === 'undefined' || !window.speechSynthesis) return;

    try {
      window.speechSynthesis.cancel();
      const utterance = new SpeechSynthesisUtterance(text);
      utterance.rate = 1.05;
      utterance.pitch = 1.05;

      const voices = window.speechSynthesis.getVoices();
      const preferredVoice = voices.find(
        v =>
          v.lang.startsWith('en') &&
          (v.name.includes('Natural') ||
            v.name.includes('Google') ||
            v.name.includes('Samantha') ||
            v.name.includes('Victoria') ||
            v.name.includes('Karen') ||
            v.name.includes('David'))
      );
      if (preferredVoice) {
        utterance.voice = preferredVoice;
      }

      utterance.onstart = () => {
        this.updateState({ isSpeaking: true });
      };
      utterance.onend = () => {
        this.updateState({ isSpeaking: false });
      };
      utterance.onerror = () => {
        this.updateState({ isSpeaking: false });
      };

      window.speechSynthesis.speak(utterance);
    } catch (err) {
      console.warn('Speech synthesis playback error:', err);
    }
  }

  public parseCommand(rawInput: string): VoiceAction {
    const cleanInput = rawInput.trim().toLowerCase();
    const normalized = cleanInput
      .replace(/^hey muse[, ]*/i, '')
      .replace(/^ok muse[, ]*/i, '')
      .replace(/^hi muse[, ]*/i, '')
      .replace(/^muse[, ]*/i, '')
      .replace(/^can you[, ]*/i, '')
      .replace(/^please[, ]*/i, '')
      .trim();

    const text = normalized.length > 0 ? normalized : cleanInput;

    // 1. Explicit Volume 0 - 100
    const volumeMatch = text.match(/(?:volume|sound|level|master)\s*(?:to|at|is|=)?\s*(\d{1,3})%?/);
    if (volumeMatch && volumeMatch[1]) {
      const vol = parseInt(volumeMatch[1], 10);
      if (!isNaN(vol)) {
        return { type: 'SET_VOLUME', targetVolume: Math.max(0, Math.min(100, vol)) };
      }
    }

    // 2. Relative Volume
    if (
      text.includes('volume down') ||
      text.includes('decrease volume') ||
      text.includes('lower volume') ||
      text.includes('turn it down') ||
      text.includes('turn down') ||
      text.includes('quieter') ||
      text.includes('softer') ||
      text.includes('reduce sound')
    ) {
      const numMatch = text.match(/(\d{1,2})/);
      const delta = numMatch ? -parseInt(numMatch[1], 10) : -15;
      return { type: 'ADJUST_VOLUME_RELATIVE', delta };
    }

    if (
      text.includes('volume up') ||
      text.includes('increase volume') ||
      text.includes('raise volume') ||
      text.includes('turn it up') ||
      text.includes('turn up') ||
      text.includes('louder') ||
      text.includes('boost volume') ||
      text.includes('higher sound')
    ) {
      const numMatch = text.match(/(\d{1,2})/);
      const delta = numMatch ? parseInt(numMatch[1], 10) : 15;
      return { type: 'ADJUST_VOLUME_RELATIVE', delta };
    }

    // 3. Mute / Unmute
    if (text === 'unmute' || text.includes('sound on') || text.includes('un-mute')) {
      return { type: 'UNMUTE' };
    }
    if (text === 'mute' || text.includes('silence') || text.includes('mute music') || text.includes('shut up')) {
      return { type: 'MUTE' };
    }

    // 4. Playback Navigation: Next, Previous, Restart
    if (
      text === 'next' ||
      text.includes('next song') ||
      text.includes('next track') ||
      text.includes('skip') ||
      text.includes('skip song') ||
      text.includes('play next') ||
      text.includes('switch song') ||
      text.includes('change song')
    ) {
      return { type: 'NEXT_SONG' };
    }

    if (
      text === 'previous' ||
      text === 'prev' ||
      text.includes('previous song') ||
      text.includes('prev song') ||
      text.includes('previous track') ||
      text.includes('go back') ||
      text.includes('last song') ||
      text.includes('play previous')
    ) {
      return { type: 'PREVIOUS_SONG' };
    }

    if (text.includes('restart') || text.includes('replay') || text.includes('start over') || text.includes('play again')) {
      return { type: 'RESTART_SONG' };
    }

    // 5. Pause / Stop
    if (
      text === 'pause' ||
      text === 'stop' ||
      text.includes('pause song') ||
      text.includes('pause music') ||
      text.includes('stop playing') ||
      text.includes('hold on') ||
      text.includes('freeze')
    ) {
      return { type: 'PAUSE' };
    }

    // 6. Play / Resume
    if (
      text === 'play' ||
      text === 'resume' ||
      text === 'continue' ||
      text === 'start' ||
      text === 'unpause' ||
      text === 'play music' ||
      text.includes('start playing') ||
      text.includes('keep playing')
    ) {
      return { type: 'PLAY' };
    }

    // 7. Muse Stream Navigation
    if (
      text.includes('open muse stream') ||
      text.includes('open stream') ||
      text.includes('muse stream') ||
      text.includes('streaming') ||
      text.includes('trending playlists') ||
      text.includes('browse stream')
    ) {
      return { type: 'OPEN_MUSE_STREAM' };
    }

    // 8. Search & Play on Muse Stream
    if (
      text.startsWith('play ') ||
      text.startsWith('stream ') ||
      text.startsWith('search and play ') ||
      text.startsWith('find and play ') ||
      text.includes('play song ') ||
      text.includes('play track ') ||
      text.includes('search for ')
    ) {
      const query = text
        .replace(/^play song /i, '')
        .replace(/^play track /i, '')
        .replace(/^play /i, '')
        .replace(/^stream /i, '')
        .replace(/^search and play /i, '')
        .replace(/^find and play /i, '')
        .replace(/^search for /i, '')
        .trim();

      if (query.length > 0) {
        return { type: 'SEARCH_AND_PLAY', query };
      }
    }

    // 9. Spatial Audio & Dolby Atmos
    if (text.includes('spatial audio') || text.includes('3d audio') || text.includes('spatial 3d')) {
      const enable = !text.includes('off') && !text.includes('disable') && !text.includes('bypass');
      return { type: 'SET_SPATIAL_AUDIO', enabled: enable };
    }
    if (text.includes('dolby atmos') || text.includes('dolby audio') || text.includes('dolby')) {
      const enable = !text.includes('off') && !text.includes('disable') && !text.includes('bypass');
      return { type: 'SET_DOLBY_ATMOS', enabled: enable };
    }

    // 10. Modals & Audio Tools
    if (text.includes('stem') || text.includes('bgm') || text.includes('isolate vocal') || text.includes('instrumental')) {
      return { type: 'OPEN_STEM_EXTRACTOR' };
    }
    if (text.includes('trim') || text.includes('crop') || text.includes('ringtone') || text.includes('alarm')) {
      return { type: 'OPEN_AUDIO_TRIMMER' };
    }
    if (text.includes('lyrics') || text.includes('words') || text.includes('meaning') || text.includes('sing along')) {
      return { type: 'OPEN_LYRICS_CHAT' };
    }

    return { type: 'GENERAL_INFO', response: `Executing voice command: "${rawInput}"` };
  }

  public parseAndExecute(rawInput: string) {
    const action = this.parseCommand(rawInput);
    const feedback = this.generateFeedback(action);

    this.updateState({
      recognizedText: rawInput,
      assistantFeedback: feedback,
      lastAction: action.type,
      isListening: false
    });

    // Speak audio feedback response aloud
    this.speak(feedback);

    if (this.onActionDispatched) {
      this.onActionDispatched(action);
    }
  }

  public generateFeedback(action: VoiceAction): string {
    switch (action.type) {
      case 'SET_VOLUME':
        return `Adjusted volume to ${action.targetVolume} percent.`;
      case 'ADJUST_VOLUME_RELATIVE':
        return action.delta > 0 ? 'Increasing volume.' : 'Lowering volume.';
      case 'MUTE':
        return 'Audio muted.';
      case 'UNMUTE':
        return 'Audio unmuted.';
      case 'NEXT_SONG':
        return 'Skipping to next track.';
      case 'PREVIOUS_SONG':
        return 'Playing previous track.';
      case 'RESTART_SONG':
        return 'Restarting track from the beginning.';
      case 'PLAY':
        return 'Resuming playback.';
      case 'PAUSE':
        return 'Playback paused.';
      case 'SEARCH_AND_PLAY':
        return `Searching and playing "${action.query}" on Muse Stream.`;
      case 'SET_SPATIAL_AUDIO':
        return action.enabled ? 'Spatial 3D Audio enabled.' : 'Spatial Audio bypassed.';
      case 'SET_DOLBY_ATMOS':
        return action.enabled ? 'Dolby Atmos 360 enabled.' : 'Dolby Atmos bypassed.';
      case 'OPEN_STEM_EXTRACTOR':
        return 'Opening BGM and Stem Studio.';
      case 'OPEN_AUDIO_TRIMMER':
        return 'Opening Ringtone and Audio Trimmer.';
      case 'OPEN_LYRICS_CHAT':
        return 'Opening Muse Lyrics Intelligence.';
      case 'OPEN_MUSE_STREAM':
        return 'Opening Muse Online Streaming Engine.';
      case 'GENERAL_INFO':
        return action.response;
    }
  }

  private updateState(partial: Partial<VoiceAssistantState>) {
    this.state = { ...this.state, ...partial };
    if (this.onStateUpdated) {
      this.onStateUpdated(this.state);
    }
  }
}
